### ISSUE TITLE
object validation creates misleading non null error

### ISSUE THREAD

**Issue by mschipperheyn** [graphql/graphql-js#1989]
# What is happening
A non null error is reported for a field that is not null, e.g.
`Cannot return null for non-nullable field Account.id.`

# What is really happening
When you dig into graphql code and start console.logging intermediate results, you start seeing very different things:

Scenario 1
`Expected a value of type "PostRegistration" but received: "TERMS"`

So, in this case, I had added a value to an ENUM field in the database but forgot to add it to the graphql enum, hence the error. Impossible to determine from the provided error message (`cannot return null ...`).

Scenario 2
`non nullable field Account.foo was null`

In this case, another required key on the same object was provided with a null value. It seems obvious, but the error message is so specific, that you can easily miss this.

In short, a whole class of field validation errors, is being aggregated to a single highly specific message that is plain wrong.

**Comment by tedb19** [comment]
any progress on this?

**Comment by djay05554** [comment]
Hello, guys

I can't find the solution to the bug.
I correctly confirm my schema but there is no wrong.

**Comment by sujit510** [comment]
I was too facing this issue while querying as follows:

**Query**
```
query schoolSession {
    students {
      id
      results {
        id
        score {
          id
          marks
          percentage
        }
      }
    }
  }
```

**Error**
```
[GraphQL error]: Message: Cannot return null for non-nullable type (line 15, column 9):
        score {
        ^, Location: [object Object], Path: schoolSession,students,9,results,0,score
```

So I realised that it is trying to fetch schoolSession.students[9].results[0].score, while schoolSession.students[9].results[0] is already been deleted. But as the relation of results to score is made mandatory in datamodel, prisma is throwing above error

So I tried by adding filter on results for non-null score, as follows:

**Query with filter**
```
query schoolSession {
    students {
      id
      results {
        id
        score {
          id
          marks
          percentage
        }
      }
    }
  }
```

This has worked for me so far.
Hope it helps someone!

**Comment by laurobd** [comment]
For me, when send a mutation using a object as input (2 levels nested, for instance), the response I have is from the last level. If I ask the _id as response, I got it from the last level:

```
type Level1 {
    name: String
    subGrp: [Level2]
}

type Level2 {
    name: String
}

input inputLevels (
    inputLevel1: String
    inputLevel2: String
)

type Mutation {
   createLevels(input: inputLevels) {

}

######

mutation createLevels($input: inputLevels) {
    createLevels(input: $input){
        _id
    }
}

```

This _id that I ask as response, brings me the "_id" of the Level 2, not the Level 1...

**Comment by timbrownsf** [comment]
Any recommendations on how to debug this if it is happening and figure out the actual error?

**Comment by mschipperheyn** [comment]
@timbrownsf Yeah, 
Easy: add console.log of the error at the error the erronous error occurs so you can see what the actual error is
Harder but better. Use server side inspect and add a breakpoint there.

**Comment by emma255** [comment]
this one works well `return Customer::findOrFail($args['id']);`
and this one not brings that error `return Customer::where('email', $args['email'])->get();`

So for myself i think there is an issue on handling exceptions in case the second query above fails, as compared to the first query above that whether find or fail.  so am continuing digging up whether i will get the solution of it..

**Comment by Nik-Novak** [comment]
I think it has to do with async functions being used when resolving. That's the common trend I'm seeing with these errors.

**Comment by vanthome** [comment]
That is really annoying, very hard to trace errors. I see something like this a good remedyhttps://github.com/graphql/graphql-js/pull/402
With that we could add proper logging for field level resolver throws.

**Comment by nidshar** [comment]
Any updates on this one?

**Comment by allaniftrue** [comment]
https://github.com/apollographql/apollo-client/issues/4180#issuecomment-493795550, this reply helped solved the issue

**Comment by evifere** [comment]
I bypass the problem using JSON.parse and JSON.stringify like this.

```TypeScript
  @Query(returns => Film, { nullable: false })
  async film(
    @Arg("episode_id")
    episode_id: number
  ) {
    let film = await FilmsModel.findOne({episode_id:episode_id});
    return JSON.parse(JSON.stringify(film)) ;
  }
```

**Comment by vipiny35** [comment]
> I bypass the problem using JSON.parse and JSON.stringify like this.
> 
> ```ts
>   @Query(returns => Film, { nullable: false })
>   async film(
>     @Arg("episode_id")
>     episode_id: number
>   ) {
>     let film = await FilmsModel.findOne({episode_id:episode_id});
>     return JSON.parse(JSON.stringify(film)) ;
>   }
> ```

Didn't work. Any other work-around? I tried using `lean()` and `toObject()` but with no luck!

**Comment by fromi** [comment]
@vipiny35 Not sure if your problem is the same, but if you want to access sub-objects functions you have to do something like this:
```
  @Query(() => Game)
  async game(@Arg('id') id: string) {
    return GameModel.findById(id)
  }

  @FieldResolver(() => GraphQLJSON, {nullable: true})
  setup(@Root() game: Game, @Ctx('userId') userId: string) {
    const player = game.players.find(player => toObjectType(Player, player).isControlledBy(userId))
    ...
  }

export default function toObjectType<T>(Type: new (...args: any[]) => T, object: T) {
  return Object.assign(new Type(), object);
}
```
You would expect "game" to be the Document from mongoose here, however in the FieldResolver it is some flat data objet, not a class anymore.

**Comment by yaacovCR** [comment]
@mschipperheyn  Can you provide a reproducible  example? (3 years later!!!)

For posterity: when we bubble up nulls, the original error is still supposed to be included in the errors array for this exact purpose. So this problem at first glance seems surprising. I may have misunderstood, of course!
